File: Baekjoon/Python/BOJ_PS/Implementation/Q3085-CandyGame.py
import sys
input = sys.stdin.readline

def count_edible(candies, n):
    max_count = 1

    for i in range(n):
        count = 1
        for j in range(1, n):
            if candies[i][j] == candies[i][j-1]:
                count += 1
            else:
                count = 1

            if max_count < count:
                max_count = count

        count = 1

        for j in range(1, n):
            if candies[j-1][i] == candies[j][i]:
                count += 1
            else:
                count = 1
            if max_count < count:
                max_count = count

    return max_count

n = int(input())
candies = [list(input()) for _ in range(n)]
result = 0

for i in range(n):
    for j in range(n):
        if j+1 < n:
            candies[i][j], candies[i][j+1] = candies[i][j+1], candies[i][j]

            max_count = count_edible(candies, n)
            if result < max_count:
                result = max_count

            candies[i][j], candies[i][j+1] = candies[i][j+1], candies[i][j]
        if i+1 < n:
            candies[i][j], candies[i+1][j] = candies[i+1][j], candies[i][j]

            max_count = count_edible(candies, n)
            if result < max_count:
                result = max_count

            candies[i][j], candies[i + 1][j] = candies[i + 1][j], candies[i][j]

print(result)

File: Baekjoon/Swift/BOJ_PS/BOJ_PS/Implementation/Q3085-CandyGame.swift
//
//  Q3085-CandyGame.swift
//  BOJ_PS
//
//  Created by thoonk on 2022/07/23.
//

func countEdible(candies: [[String]], n: Int) -> Int {
    var maxCount = 1
    
    for i in 0..<n {
        var count = 1
        for j in 1..<n {
            if candies[i][j-1] == candies[i][j] {
                count += 1
            } else {
                count = 1
            }
            
            if count > maxCount {
                maxCount = count
            }
        }
        
        count = 1
        for j in 1..<n {
            if candies[j-1][i] == candies[j][i] {
                count += 1
            } else {
                count = 1
            }
            
            if count > maxCount {
                maxCount = count
            }
        }
    }
    
    return maxCount
}

let n = Int(readLine()!)!
var candies = [[String]]()
var result = 0

for _ in 0..<n {
    let candy = Array(readLine()!.map { String($0) })
    candies.append(candy)
}

for i in 0..<n {
    for j in 0..<n {
        if j+1 < n {
            candies[i].swapAt(j, j+1)
            
            let maxCount = countEdible(candies: candies, n: n)
            if maxCount > result {
                result = maxCount
            }
            
            candies[i].swapAt(j, j+1)
        }
        
        if i+1 < n {
            var temp = candies[i][j]
            candies[i][j] = candies[i+1][j]
            candies[i+1][j] = temp
            
            let maxCount = countEdible(candies: candies, n: n)
            if maxCount > result {
                result = maxCount
            }
            
            temp = candies[i+1][j]
            candies[i+1][j] = candies[i][j]
            candies[i][j] = temp
        }
    }
}

print(result)


